diesel/src/pg/connection/mod.rs

```diff
@@ -134,51 +244,163 @@ impl PgConnection {
-            });
+    fn accepts(ty: &Type) -> bool {
+        ty.oid() != 0
+    }
 
-        Ok((query?, binds))
+    fn to_sql_checked(
+        &self,
+        ty: &Type,
+        out: &mut BytesMut,
+    ) -> Result<postgresql::types::IsNull, Box<dyn Error + 'static + Send + Sync>> {
+        if !Self::accepts(ty) {
+            todo!("write a proper error message")
+        }
+        self.to_sql(ty, out)
     }
+}
 
-    fn execute_inner(&self, query: &str) -> QueryResult<PgResult> {
-        let query = Statement::prepare(self, query, None, &[])?;
-        query.execute(self, &Vec::new())
+struct PostgresRow {
+    row: postgresql::row::Row,
+    col_idx: usize,
+}
+
+impl Row<Pg> for PostgresRow {
+    fn take(&mut self) -> Option<PgValue> {
+        let current_index = self.col_idx;
+        self.col_idx += 1;
+        let DieselFromSqlWrapper(value) = self.row.get(current_index);
+        value
     }
 
-    fn set_config_options(&self) -> QueryResult<()> {
-        self.execute("SET TIME ZONE 'UTC'")?;
-        self.execute("SET CLIENT_ENCODING TO 'UTF8'")?;
-        self.raw_connection
-            .set_notice_processor(noop_notice_processor);
-        Ok(())
+    fn next_is_null(&self, count: usize) -> bool {
+        (0..count).all(|i| {
+            let DieselFromSqlWrapper(value) = self.row.get(self.col_idx + i);
+            value.is_none()
+        })
     }
+}
 
-    pub(crate) fn get_metadata_cache(&self) -> &PgMetadataCache {
-        &self.metadata_cache
+struct DieselFromSqlWrapper<'a>(Option<PgValue<'a>>);
+
+impl<'a> postgresql::types::FromSql<'a> for DieselFromSqlWrapper<'a> {
+    fn from_sql(ty: &Type, raw: &'a [u8]) -> Result<Self, Box<dyn Error + 'static + Send + Sync>> {
+        use std::num::NonZeroU32;
+
+        Ok(DieselFromSqlWrapper(Some(PgValue::new(
+            raw,
+            NonZeroU32::new(ty.oid()).expect("That's not 0"),
+        ))))
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        ty.oid() != 0
+    }
+
+    fn from_sql_null(_ty: &Type) -> Result<Self, Box<dyn Error + Sync + Send>> {
+        Ok(DieselFromSqlWrapper(None))
+    }
+}
+
+struct NamedPostgresRow {
+    row: postgresql::row::Row,
+}
+
+impl NamedRow<Pg> for NamedPostgresRow {
+    fn index_of(&self, column_name: &str) -> Option<usize> {
+        self.row
+            .columns()
+            .iter()
+            .position(|c| c.name() == column_name)
+    }
+
+    fn get_raw_value(&self, index: usize) -> Option<PgValue> {
+        let DieselFromSqlWrapper(value) = self.row.get(index);
+        value
     }
 }
 
-extern "C" fn noop_notice_processor(_: *mut libc::c_void, _message: *const libc::c_char) {}
+impl From<postgresql::error::Error> for crate::result::Error {
+    fn from(e: postgresql::error::Error) -> crate::result::Error {
+        use crate::result::DatabaseErrorKind::*;
+        use postgresql::error::SqlState;
+
+        if let Some(code) = e.code() {
+            let kind = if *code == SqlState::UNIQUE_VIOLATION {
```

Comment:
should probably implement the From trait in a different file for these
I added the missing mappings

Reply:
I've put everything in one file back then, because that made development easier. If you find a good location for this, that would be great.